docs/developing/auth.md

```diff
@@ -95,15 +95,49 @@ the
 [sandstorm-rawapi-example](https://github.com/sandstorm-io/sandstorm-rawapi-example)
 repository on GitHub.
 
-## Further reading
 
-You might be interested in looking at:
+## Defining permissions and roles
+
+Apps define permissions by providing a
+[`UiView.ViewInfo`](https://github.com/sandstorm-io/sandstorm/blob/v0.177/src/sandstorm/grain.capnp#L160-L265) to Sandstorm.
+Apps that use `sandstorm-http-bridge`
+can specify a `ViewInfo` value in the `PackageDefinition.bridgeConfig` field
+of their `sandstorm-pkgdef.capnp`.
+Apps that do not use `sandstorm-http-bridge`
+can directly provide a `ViewInfo` by implementing the
+[`UiView.getViewInfo()`](https://github.com/sandstorm-io/sandstorm/blob/v0.177/src/sandstorm/grain.capnp#L157) Cap'n Proto method.
+
+When sharing a grain, a user selects a bundle of permissions to grant.
+Such a bundle of permissions is called a *role*.
+Roles are intended to give users a more human-friendly
+handle on permissions and to steer users away from
+combinitions of permissions that might not make sense.
+Like permissions, roles are defined in an app's `ViewInfo`.
+
+From Sandstorm's perspective, the meanings of permissions are completely opaque.
+Sandstorm merely tracks who is allowed to access which grain with which permissions.
+Sandstorm represents those permissions as a bit vector and leaves it up to the app
+to interpret those bits in an appropriate way.
+When a share takes place, Sandstorm records the *role* that was shared, but not the
+precise *permissions*, which are are computed on-the-fly every time the recipient
+of the share opens the grain.
+Therefore, if a later version of the app modifies the role definition,
+existing shares will be affected.
+
+In a `ViewInfo`, permissions and roles are defined in lists of
+[`PermissionDef`](https://github.com/sandstorm-io/sandstorm/blob/v0.177/src/sandstorm/grain.capnp#L524-L545) and
+[`RoleDef`](https://github.com/sandstorm-io/sandstorm/blob/v0.177/src/sandstorm/grain.capnp#L547-L579)
+values. Later versions of an app can always add more permissions or
+roles, but it is important to never remove any element from these lists.
+Instead, you can set the `obsolete` field to `true`.
+
+Here are some examples:
 
 * A [sandstorm-pkgdef.capnp](https://github.com/kentonv/ssjekyll/blob/fd09dbdbd6644abe63c50060044b71556130c30d/sandstorm-pkgdef.capnp)
   with no permissions defined.
 
 * A [sandstorm-pkgdef.capnp](https://github.com/jparyani/mediawiki-sandstorm/blob/8c7a7d10b6121cb5e94247f7ea27a46ebf8e84eb/sandstorm-pkgdef.capnp)
   with one permission defined.
 
-* The [implementation of
-  sandstorm-http-bridge](https://github.com/sandstorm-io/sandstorm/blob/411b344f3acb151693036f3c061b153a2fd91d68/src/sandstorm/sandstorm-http-bridge.c%2B%2B).
+* A [sandstorm-pkgdef.capnp](https://github.com/dwrensha/groovebasin/blob/c6a2cbda0b7a94971f9671a6b4955e1007470556/.sandstorm/sandstorm-pkgdef.capnp)
+  with five permissions defined.
```

Comment:
I suppose it would be nice to include a sample capnp which does _not_ depend on `sandstorm-http-bridge`, if one is available!